apps/messages/src/template-plugin.ts

```diff
@@ -32,31 +36,54 @@ const htmlTemplate = (
 </body>
 </html>`;
 
-const apply = async (stats, origin) => {
+const apply = async (stats: Stats) => {
   const outputDir = stats.compilation.outputOptions.path;
+  if (!outputDir) {
+    throw new Error('Cannot determine output dir');
+  }
+
   const files = await fs.readdir(outputDir);
   const tasks = files
-    .filter((f) => f.endsWith('.js'))
-    .map(async (f) => {
+    .filter((file) => path.extname(file) === '.js')
+    .map(async (file) => {
       const cache = createCache();
       const { extractCritical } = createEmotionServer(cache);
-      const pkg = require(path.resolve(outputDir, f));
-      const Component = pkg.default;
+      // eslint-disable-next-line
+      const template = require(path.resolve(outputDir, file)) as Record<
+        string,
+        unknown
+      >;
+      const { name } = path.parse(file);
 
-      const { name } = path.parse(f);
-      const subject = pkg.subject ? pkg.subject : titleCase(name);
+      const reactElement = template.default as React.ReactElement;
+      if (
+        typeof reactElement === 'object' &&
+        !React.isValidElement(reactElement)
+      ) {
+        throw new Error(
+          `Template ${file} does not default export a valid React element`,
+        );
+      }
+
+      const subject = template.subject as string | undefined;
+      if (!(typeof subject === 'string')) {
+        throw new Error(`Template ${file} does not export a subject string`);
+      }
 
       const element = React.createElement(
         CacheProvider,
         { value: cache },
-        React.createElement(Component),
+        reactElement,
       );
 
       const { html, css } = extractCritical(renderToStaticMarkup(element));
       const s3Html = Object.keys(stats.compilation.assets).reduce(
         (res, asset) => {
           const { base } = path.parse(asset);
-          return res.replace(asset, `${origin}/static/media/${base}`);
+          return res.replace(
+            asset,
+            new URL(`/.messages-static/${base}`, APP_ORIGIN).toString(),
```

Comment:
This was perhaps the biggest architecture problem - images in the email were linking to assets deployed as part of the frontend, hoping that create-react-app constructed the image path with the `[name][hash:8][ext]` etc in the same way. I made a separately deployed bucket for the static assets in the emails now

Reply:
Maybe we could inline the images instead?

Reply:
```
Some bad news about base64 encoded images:

totally blocked by Outlook
not displayed by most webmail services (especially if you use more than one inside the message).
```

Reply:
also would be worse for perf since they aren't lazy loaded.

Reply:
So I get why Filipe did it this way, it was just missing a tiny bit to be reliable and not rely on what the FE deploys :D